pyfive/dataobjects.py

```diff
@@ -482,74 +487,131 @@ def find_msg_type(self, msg_type):
-                link_name_character_set = struct.unpack_from('<B', self.msg_data, offset)[0]
-                offset += 1
-            else:
-                link_name_character_set = 0
+            yield from table.get_links(heap).items()
+
+    def _get_link_from_link_msg(self, link_msg):
+        """ Retrieve link from link message. """
+        offset = link_msg['offset_to_message']
+        return self._decode_link_msg(self.msg_data, offset)[1]
+
+    @staticmethod
+    def _decode_link_msg(data, offset, dereference=True):
+        version, flags = struct.unpack_from('<BB', data, offset)
+        offset += 2
+        assert version == 1
+
+        size_of_length_of_link_name = 2**(flags & 3)
+        link_type_field_present = flags & 2**3
+        link_name_character_set_field_present = flags & 2**4
+        ordered = flags & 2**2
+
+        if link_type_field_present:
+            link_type = struct.unpack_from('<B', data, offset)[0]
+            offset += 1
+        else:
+            link_type = 0
+        assert link_type in [0,1]
 
-            encoding = 'ascii' if link_name_character_set == 0 else 'utf-8'
+        if ordered:
+            creationorder = struct.unpack_from('<Q', data, offset)[0]
+            offset += 8
+        else:
+            creationorder = None
+
+        if link_name_character_set_field_present:
+            link_name_character_set = struct.unpack_from('<B', data, offset)[0]
+            offset += 1
+        else:
+            link_name_character_set = 0
+
+        encoding = 'ascii' if link_name_character_set == 0 else 'utf-8'
 
-            name_size_fmt = ["<B", "<H", "<I", "<Q"][flags & 3]
-            name_size = struct.unpack_from(name_size_fmt, self.msg_data, offset)[0]
-            offset += size_of_length_of_link_name
-            name = self.msg_data[offset:offset+name_size].decode(encoding)
-            offset += name_size
+        name_size_fmt = ["<B", "<H", "<I", "<Q"][flags & 3]
+        name_size = struct.unpack_from(name_size_fmt, data, offset)[0]
+        offset += size_of_length_of_link_name
 
+        name = data[offset:offset+name_size].decode(encoding)
+        offset += name_size
+
+        if dereference:
             if link_type == 0:
                 # hard link
-                address = struct.unpack_from('<Q', self.msg_data, offset)[0]
-                links[name] = address
+                address = struct.unpack_from('<Q', data, offset)[0]
             elif link_type == 1:
                 # soft link
-                length_of_soft_link_value = struct.unpack_from('<H', self.msg_data, offset)[0]
+                length_of_soft_link_value = struct.unpack_from('<H', data, offset)[0]
                 offset += 2
-                links[name] = self.msg_data[offset:offset+length_of_soft_link_value].decode(encoding)
-        return links
+                address = data[offset:offset+length_of_soft_link_value].decode(encoding)
+
+        return creationorder, (name, address)
```

Comment:
address will be undefined if dereference == False

Reply:
Thanks, I'll remove it. Lets keep optional dereferencing for another PR.

Reply:
Done